spices/SPICE-0016-http-rewrites.adoc

```diff
@@ -0,0 +1,310 @@
+
+Sample message:
+
+[source]
+----
+Pkl Error
+An I/O error occurred when making request GET /foo:
+Connection reset (request was rewritten: https://foo.com -> https://bar.com)
+----
+
+[[configuration-api]]
+=== Configuration API
+
+HTTP rewrites can be configured via the standard library, in Java as an API, as a server message API, and as a Gradle API.
+
+Users of the `pkl` CLI are encouraged to configure HTTP rewrites in `~/.pkl/settings.pkl` for the purpose of package mirroring.
+This will have the effect of enabling mirroring for all `pkl` CLI calls for that shell user.
+
+==== Standard Library
+
+A new property will be added to class `pkl.EvaluatorSettings.Http`.
+
+.EvaluatorSettings.pkl
+[source,pkl]
+----
+class Http {
+  @Since { version = "0.29.0" }
+  rewrites: Mapping<Uri, Uri(startsWith("http://") || startsWith("https://"))>?
+}
+----
+
+This property can then be set in a PklProject, or the settings file, because they both have properties of type `Http`:
+
+.PklProject
+[source,pkl]
+----
+amends "pkl:PklProject"
+
+evaluatorSettings {
+  http {
+    rewrites {
+      // etc
+    }
+  }
+}
+----
+
+~/.pkl/settings.pkl
+[source,pkl]
+----
+amends "pkl:settings"
+
+http {
+  rewrites {
+    // etc
+  }
+}
+----
+
+==== Java API
+
+Users of the Java API can configure HTTP settings when building `org.pkl.core.http.HttpClient` via the builder.
+
+The following methods will be added:
+
+[source,java]
+----
+interface HttpClient {
+  interface Builder {
+    Builder setRewrites(Map<String, String> rewrites);
+
+    Builder addRewrite(String sourcePrefix, String targetPrefix);
+  }
+}
+----
+
+==== Server API
+
+Users of other language bindings (Kotlin, Go, etc), can configure rewrites via property `http` in `CreateEvaluatorRequest`:
+
+[source,pkl]
+----
+class CreateEvaluatorRequest {
+  http: Http
+}
+
+class Http {
+  rewrites: Mapping<String, String>?
+}
+----
+
+==== Gradle API
+
+Class `BasePklSpec` will have a new property:
+
+[source,java]
+----
+class BasePklSpec {
+  MapProperty<String, String> getHttpRewrites();
+}
+----
+
+It can then be set in any Pkl task:
+
+.build.gradle.kts
+[source,kotlin]
+----
+pkl {
+  evaluators {
+    register("evalPkl") {
+      httpRewrites = mapOf("https://foo.com" to "https://bar.com")
+    }
+  }
+}
+----
+
+==== CLI flag
+
+A new flag is introduced: `--http-rewrite`.
+
+It accepts `key=value` pairs, and can be repeated.
+
+Example:
+[source,shell]
+----
+pkl eval \
+  --http-rewrite https://foo.com=https://bar.com \
+  myModule.pkl
+----
+
+This flag will be available to commands that use common CLI options (for example, `eval`, `test`, `project`).
+
+=== Security checks
+
+All outbound requests will be checked against the `--allowed-modules` and `--allowed-resources` CLI flags, and their respective settings in other contexts.
+This is also the case with URL rewriting.
+
+=== Error modes
+
+==== Invalid scheme
+
+An HTTP target URL must start with `http://` or `https://`.
```

Comment:
I wonder if it makes sense to prevent rewriting HTTPS URIs to HTTP URIs for security URIs. This seems like a potential "downgrade" sort of attack surface.

Reply:
I've considered this, but I don't think this is necessary.

If users want to disable HTTP schemes, the correct path for that is to use the `--allowed-modules` and `--allowed-resources` flags.